subsys/llext/llext_link.c

```diff
@@ -188,39 +188,44 @@ int llext_link(struct llext_loader *ldr, struct llext *ext, bool do_local)
 			continue;
 		}
 
+		if (shdr->sh_info >= ldr->sect_cnt ||
+		    shdr->sh_size % shdr->sh_entsize != 0) {
+			LOG_ERR("Sanity checks failed for section %d "
+				"(info %zd, size %zd, entsize %zd)", i,
+				(size_t)shdr->sh_info,
+				(size_t)shdr->sh_size,
+				(size_t)shdr->sh_entsize);
+			return -EINVAL;
+		}
+
 		rel_cnt = shdr->sh_size / shdr->sh_entsize;
 
 		name = llext_string(ldr, ext, LLEXT_MEM_SHSTRTAB, shdr->sh_name);
 
-		if (strcmp(name, ".rel.text") == 0) {
-			loc = (uintptr_t)ext->mem[LLEXT_MEM_TEXT];
-		} else if (strcmp(name, ".rel.bss") == 0 ||
-			   strcmp(name, ".rela.bss") == 0) {
-			loc = (uintptr_t)ext->mem[LLEXT_MEM_BSS];
-		} else if (strcmp(name, ".rel.rodata") == 0 ||
-			   strcmp(name, ".rela.rodata") == 0) {
-			loc = (uintptr_t)ext->mem[LLEXT_MEM_RODATA];
-		} else if (strcmp(name, ".rel.data") == 0) {
-			loc = (uintptr_t)ext->mem[LLEXT_MEM_DATA];
-		} else if (strcmp(name, ".rel.exported_sym") == 0) {
-			loc = (uintptr_t)ext->mem[LLEXT_MEM_EXPORT];
-		} else if (strcmp(name, ".rela.plt") == 0 ||
-			   strcmp(name, ".rela.dyn") == 0) {
-			llext_link_plt(ldr, ext, shdr, do_local, NULL);
-			continue;
-		} else if (strncmp(name, ".rela", 5) == 0 && strlen(name) > 5) {
-			elf_shdr_t *tgt = llext_section_by_name(ldr, name + 5);
+		/*
+		 * FIXME: The Xtensa port is currently using a different way of
+		 * handling relocations that ultimately results in separate
+		 * arch-specific code paths. This code should be merged with
+		 * the logic below once the differences are resolved.
+		 */
+		if (IS_ENABLED(CONFIG_XTENSA)) {
```

Comment:
Please review the associated commit description. That is not Xtensa specific because of the naming, it's because if you enter `link_plt` you need to ultimately implement a _different_ architecture-specific relocation function (`arch_elf_relocate_local`), and that is defined only on Xtensa. 
Furthermore, as the previous comments re/SOF imply, there are subtle undocumented issues in that code path that make it brittle.

Reply:
We need to unify the relocate arch api imo, plt or pie shouldn’t require a different entry